src/network/ClientConnectionManager.ts

```diff
@@ -0,0 +1,804 @@
+    }
+
+    private connectToCluster(): Promise<void> {
+        if (this.asyncStart) {
+            return this.submitConnectToClusterTask();
+        } else {
+            return this.doConnectToCluster();
+        }
+    }
+
+    private submitConnectToClusterTask(): Promise<void> {
+        if (this.connectToClusterTaskSubmitted) {
+            return Promise.resolve();
+        }
+
+        this.doConnectToCluster()
+            .then(() => {
+                this.connectToClusterTaskSubmitted = false;
+                if (this.activeConnections.size === 0) {
+                    this.logger.warn('ConnectionManager', 'No connection to cluster ' + this.clusterId);
+                    return this.submitConnectToClusterTask();
+                }
+            })
+            .catch((error: Error) => {
+                this.logger.warn('ConnectionManager', 'Could not connect to any cluster, shutting down ' +
+                    'the client: ' + error.message);
+                this.shutdownClient();
+            });
+
+        this.connectToClusterTaskSubmitted = true;
+        return Promise.resolve();
+    }
+
+    private doConnectToCluster(): Promise<void> {
+        const triedAddresses = new Set<string>();
+        this.waitStrategy.reset();
+        return this.tryConnectingToAddresses(triedAddresses)
+            .then((isConnected) => {
+                if (isConnected) {
+                    return;
+                }
+                this.logger.info('ConnectionManager', 'Unable to connect any address from the cluster ' +
+                    'with the name: ' + this.client.getConfig().clusterName +
+                    '. The following addresses were tried: ' +
+                    Array.from(triedAddresses));
+
+                const message = this.client.getLifecycleService().isRunning()
+                    ? 'Unable to connect any cluster.' : 'Client is being shutdown.';
+                throw new IllegalStateError(message);
+            });
+    }
+
+    private tryConnectingToAddresses(triedAddresses: Set<string>): Promise<boolean> {
+        return this.getPossibleMemberAddresses()
+            .then((addresses) => {
+                return this.tryConnectingToAddress(0, addresses, triedAddresses);
+            })
+            .then((isConnected) => {
+                if (isConnected) {
+                    return true;
+                }
+                // If the address providers load no addresses (which seems to be possible),
+                // then the above loop is not entered and the lifecycle check is missing,
+                // hence we need to repeat the same check at this point.
+                if (!this.client.getLifecycleService().isRunning()) {
+                    return Promise.reject(new ClientNotActiveError('Client is not active.'));
+                }
+                return this.waitStrategy.sleep()
+                    .then((notTimedOut) => {
```

Comment:
nit: nested `then()` calls should be avoided, when possible, as this helps readability.

Reply:
If I move the then block, then I need to add a check for `                if (isConnected) {
                    return true;
                }` again. Should I do it ?

Reply:
You're right. The `if` block complicates things here, so it's better to keep nested `.then()` calls, as they're now.